File: src/api/ShowNTell.Domain/Models/ImagePost.cs
﻿using System;
using System.Collections.Generic;
using System.Text;

namespace ShowNTell.Domain.Models
{
    /// <summary>
    /// A model for an image post.
    /// </summary>
    public class ImagePost
    {
        public int Id { get; set; }
        public string UserEmail { get; set; }
        public string ImageUri { get; set; }
        public string Description { get; set; }
        public DateTime DateCreated { get; set; }

        public User User { get; set; }
        public ICollection<ImagePostTag> Tags { get;set; }
        public ICollection<Like> Likes { get;set; }
        public ICollection<Comment> Comments { get;set; }
    }
}

File: src/show-n-tell-vue/src/models/image-post.js
import User from './user'
import Like from './like'
import Comment from './comment'

function ImagePost(id, userEmail, imageUri, description, dateCreated, tags, user, likes, comments) {
    this.id = id
    this.userEmail = userEmail
    this.imageUri = imageUri
    this.description = description
    this.dateCreated = new Date(dateCreated)

    this.tags = tags
    this.user = user
    this.likes = likes
    this.comments = comments

    this.username = this.user.username
}

ImagePost.fromJSON = function(imagePost) {
    return new ImagePost(
        imagePost.id, 
        imagePost.userEmail, 
        imagePost.imageUri, 
        imagePost.description, 
        imagePost.dateCreated, 
        imagePost.tags ? imagePost.tags.map(t => t.content) : [],
        imagePost.user ? User.fromJSON(imagePost.user) : {}, 
        imagePost.likes ? imagePost.likes.map(l => Like.fromJSON(l)) : [], 
        imagePost.comments ? imagePost.comments.map(c => Comment.fromJSON(c)) : [])
}

export default ImagePost